src/libraries/Common/src/System/Reflection/Metadata/TypeNameParserHelpers.cs

```diff
@@ -0,0 +1,377 @@
+                {
+                    // do not validate the type name now, it will be validated as a whole nested type name later
+                    (nestedNameLengths ??= new()).Add(length);
+                    totalLength += 1; // skip the '+' sign in next search
+                }
+                totalLength += length;
+            } while (isNestedType);
+
+            return true;
+        }
+
+        internal static bool TryParseNextDecorator(ref ReadOnlySpan<char> input, out int rankOrModifier)
+        {
+            // Then try pulling a single decorator.
+            // Whitespace cannot precede the decorator, but it can follow the decorator.
+
+            ReadOnlySpan<char> originalInput = input; // so we can restore on 'false' return
+
+            if (TryStripFirstCharAndTrailingSpaces(ref input, '*'))
+            {
+                rankOrModifier = Pointer;
+                return true;
+            }
+
+            if (TryStripFirstCharAndTrailingSpaces(ref input, '&'))
+            {
+                rankOrModifier = ByRef;
+                return true;
+            }
+
+            if (TryStripFirstCharAndTrailingSpaces(ref input, '['))
+            {
+                // SZArray := []
+                // MDArray := [*] or [,] or [,,,, ...]
+
+                int rank = 1;
+                bool hasSeenAsterisk = false;
+
+            ReadNextArrayToken:
+
+                if (TryStripFirstCharAndTrailingSpaces(ref input, ']'))
+                {
+                    // End of array marker
+                    rankOrModifier = rank == 1 && !hasSeenAsterisk ? SZArray : rank;
+                    return true;
+                }
+
+                if (!hasSeenAsterisk)
+                {
+                    if (rank == 1 && TryStripFirstCharAndTrailingSpaces(ref input, '*'))
+                    {
+                        // [*]
+                        hasSeenAsterisk = true;
+                        goto ReadNextArrayToken;
+                    }
+                    else if (TryStripFirstCharAndTrailingSpaces(ref input, ','))
+                    {
+                        // [,,, ...]
+                        checked { rank++; }
+                        goto ReadNextArrayToken;
+                    }
+                }
+
+                // Don't know what this token is.
+                // Fall through to 'return false' statement.
+            }
+
+            input = originalInput; // ensure 'ref input' not mutated
+            rankOrModifier = 0;
+            return false;
+        }
+
+        internal static bool TryStripFirstCharAndTrailingSpaces(ref ReadOnlySpan<char> span, char value)
+        {
+            if (!span.IsEmpty && span[0] == value)
+            {
+                span = span.Slice(1).TrimStart();
+                return true;
+            }
+            return false;
+        }
+
+        internal static InvalidOperationException InvalidOperation_MaxNodesExceeded(int limit) =>
```

Comment:
It would be better for these methods to actually throw instead of returning the exception. It is the preferred ThrowHelper pattern going forward.